### PR TITLE
Use new Fluo and Fluo Recipes APIs

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

**Comment by keith-turner** [comment]
This depends on apache/incubator-fluo-recipes#128

File: modules/data/src/main/java/webindex/data/fluo/WebindexObservers.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  @Override
+  public void provide(Registry obsRegistry, Context ctx) {
+    SimpleConfiguration appCfg = ctx.getAppConfiguration();
+    MetricsReporter reporter = ctx.getMetricsReporter();
+
+    // Create an export queue that handles all updates to the query table.
+    ExportQueue<String, IndexUpdate> exportQ =
+        ExportQueue.getInstance(FluoApp.EXPORT_QUEUE_ID, appCfg);
+
+    // Create a CFM that tracks the number of pages linking to a URI.
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/WebindexObservers.java]: should be `CQ`

File: modules/data/src/main/java/webindex/data/fluo/DomainMap.java [hunk 2]
```diff
@@ -14,69 +14,38 @@
...
-import org.apache.fluo.recipes.core.map.CollisionFreeMap;
-import org.apache.fluo.recipes.core.map.CollisionFreeMap.Options;
-import org.apache.fluo.recipes.core.map.Combiner;
-import org.apache.fluo.recipes.core.map.Update;
-import org.apache.fluo.recipes.core.map.UpdateObserver;
 import webindex.core.models.export.DomainUpdate;
 import webindex.core.models.export.IndexUpdate;
-import webindex.data.FluoApp;
 
 public class DomainMap {
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/DomainMap.java]: `DomainMap` could be renamed to `DomainQueue` or `DomainCombiner`

File: modules/data/src/main/java/webindex/data/fluo/PageObserver.java [hunk 3]
```diff
@@ -40,14 +41,13 @@
...
-import webindex.data.FluoApp;
 
-public class PageObserver extends AbstractObserver {
+public class PageObserver implements Observer {
 
   private static final Logger log = LoggerFactory.getLogger(PageObserver.class);
   private static final Gson gson = new Gson();
 
-  private CollisionFreeMap<String, UriInfo> uriMap;
+  private CombineQueue<String, UriInfo> uriMap;
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/PageObserver.java]: `uriMap` could be renamed to `uriQ` or `uriCombiner`

File: modules/data/src/main/java/webindex/data/fluo/WebindexObservers.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    CombineQueue<String, UriInfo> uriMap = CombineQueue.getInstance(UriMap.URI_MAP_ID, appCfg);
+
+    // Create a CFM that tracks the number of unique URIs observed per domain.
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/WebindexObservers.java]: should be `CQ`

File: docs/code-guide.md [hunk 5]
```diff
@@ -21,64 +21,61 @@ The image was produced using Google Docs.  A link to the source is here.
...
-simple because it builds on the Collision Free Map Recipe. A Collision Free Map has two extension
-points and this example implements both. The first extension point is a combiner that processes the
-`+1` and `-1` updates queued up by the [PageObserver]. The second extension point is an update
-observer that handles changes in reference counts for a URI.  It pushes these changes in reference
-counts to the [DomainMap] and [IndexExporter].
+### URI Map 
 
-Changes to URI reference counts are aggregated per domain and `+1` and `-1` updates are queued for
-the [DomainMap].
+A Collision Free Map is setup to track the number of pages linking to a URI.  The `reduce()` function
```
Comment (mikewalch) [inline docs/code-guide.md]: A CombineQueue is

File: docs/code-guide.md [hunk 6]
```diff
@@ -21,64 +21,61 @@ The image was produced using Google Docs.  A link to the source is here.
...
+observer also queues change in URI inbound link counts to the export queue.
 
 ### Domain Map Observer
 
-The [DomainMap] observer computers per domain reference counts. This is a Collision Free Map that
-tracks per domain information. When its notified that domain counts changed, it queues updates for
-the [IndexExporter] to update the Query table.
-
-### IndexExporter
+A Collision Free Map is setup to track the number of unique URIs observed in each domain.  The
```
Comment (mikewalch) [inline docs/code-guide.md]: A CombineQueue

File: modules/data/src/main/java/webindex/data/fluo/UriMap.java [hunk 7]
```diff
@@ -15,25 +15,19 @@
...
-import org.apache.fluo.recipes.core.map.Update;
-import org.apache.fluo.recipes.core.map.UpdateObserver;
 import webindex.core.models.URL;
 import webindex.core.models.UriInfo;
 import webindex.core.models.export.IndexUpdate;
 import webindex.core.models.export.UriUpdate;
-import webindex.data.FluoApp;
 
 /**
  * This class contains code related to a CollisionFreeMap that keeps track of the count of
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/UriMap.java]: should be CombineQueue

File: modules/data/src/main/java/webindex/data/fluo/UriCombineQ.java [hunk 8]
```diff
@@ -15,83 +15,52 @@
...
 /**
- * This class contains code related to a CollisionFreeMap that keeps track of the count of
- * information about URIs.
+ * This class contains code related to a CombineQueue that keeps track of the count of information
+ * about URIs.
  */
-public class UriMap {
+public class UriCombineQ {
 
   public static final String URI_MAP_ID = "um";
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/UriCombineQ.java]: This should be renamed

File: modules/data/src/main/java/webindex/data/fluo/UriCombineQ.java [hunk 9]
```diff
@@ -15,83 +15,52 @@
...
-
-  /**
    * Observes uri map updates and adds those updates to an export queue.
    */
-  public static class UriUpdateObserver extends UpdateObserver<String, UriInfo> {
+  public static class UriUpdateObserver implements ChangeObserver<String, UriInfo> {
 
     private ExportQueue<String, IndexUpdate> exportQ;
-    private CollisionFreeMap<String, Long> domainMap;
+    private CombineQueue<String, Long> domainMap;
```
Comment (mikewalch) [inline modules/data/src/main/java/webindex/data/fluo/UriCombineQ.java]: could be `domainQ`

File: docs/code-guide.md [hunk 10]
```diff
@@ -21,64 +21,63 @@ The image was produced using Google Docs.  A link to the source is here.
...
+### Page Observer
 
-The [UriMap] observer computes per URI reference counts. The code for this this observer is very
-simple because it builds on the Collision Free Map Recipe. A Collision Free Map has two extension
-points and this example implements both. The first extension point is a combiner that processes the
-`+1` and `-1` updates queued up by the [PageObserver]. The second extension point is an update
-observer that handles changes in reference counts for a URI.  It pushes these changes in reference
-counts to the [DomainMap] and [IndexExporter].
+The [PageObserver] computes changes to links in a page.  It queues `+1` and `-1`  for  new and
+deleted URIs to the uriMap.  It also queues up changes in URIs to the export queue.
```
Comment (mikewalch) [inline docs/code-guide.md]: should be `uriQ`

**Pull request merged**